baseapp/abci.go

```diff
@@ -199,10 +201,24 @@ func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeg
 	// set the signed validators for addition to context in deliverTx
 	app.voteInfos = req.LastCommitInfo.GetVotes()
 
-	// call the hooks with the BeginBlock messages
-	for _, streamingListener := range app.abciListeners {
-		if err := streamingListener.ListenBeginBlock(app.deliverState.ctx, req, res); err != nil {
-			panic(fmt.Errorf("BeginBlock listening hook failed, height: %d, err: %w", req.Header.Height, err))
+	// call the streaming service hook with the BeginBlock messages
+	for _, abciListener := range app.abciListeners {
+		ctx := app.deliverState.ctx
+		blockHeight := ctx.BlockHeight()
+		if app.abciListenersAsync {
+			go func(req abci.RequestBeginBlock, res abci.ResponseBeginBlock) {
+				if err := abciListener.ListenBeginBlock(ctx, req, res); err != nil {
+					app.logger.Error("BeginBlock listening hook failed", "height", blockHeight, "err", err)
+				}
+			}(req, res)
+		} else {
+			if err := abciListener.ListenBeginBlock(ctx, req, res); err != nil {
+				app.logger.Error("BeginBlock listening hook failed", "height", blockHeight, "err", err)
+				if app.stopNodeOnABCIListenerErr {
+					plugin.CleanupClients()
+					os.Exit(1)
```

Comment:
I put this here for feedback because I don't have an idea yet where it should go. I don't like it that the API is aware of the plugin cleanup here. Perhaps fold it into a function in the BaseApp? Thoughts?

=> `app.stopABCIListenersAndExit()`?


@tac0turtle @alexanderbez @peterbourgon

Reply:
Could the `abci.ListenBeginBlock` implementation not handle both `os.Exit(1)` and cleanup? This would simplify the top level control flow here.

Reply:
What are your thoughts @egaxhaj? Seems like pushing this down to `ListenBeginBlock` is cleaner, no?

Reply:
Moved down to the plugin.

https://github.com/cosmos/cosmos-sdk/blob/31a9e9c49299bbc65f804d3bf99182da98f43c83/store/streaming/abci/grpc.go#L88